new-log-viewer/src/services/LogFileManager.ts

```diff
@@ -0,0 +1,138 @@
+
+
+/**
+ * A mapping of file name extensions to their respective decoders.
+ */
+const FILE_EXT_TO_DECODER: Readonly<Record<string, DecoderConstructor>> = Object.freeze({
+    ".jsonl": JsonlDecoder as unknown as DecoderConstructor,
+});
+
+class LogFileManager {
+    #pageSize: number;
+
+    #fileData: Uint8Array | null = null;
+
+    #fileName: string | null = null;
+
+    #numEvents: number = 0;
+
+    #decoder: Decoder | null = null;
+
+
+    /**
+     * Retrieves the decoder constructor based on the file extension.
+     *
+     * @param filename The filename from which to extract the file extension.
+     * @return The constructor associated with the file extension, or null if not found.
+     */
+    static #getDecoderConstructor = (filename: string): DecoderConstructor | null => {
+        for (const ext in FILE_EXT_TO_DECODER) {
+            if (Object.hasOwn(FILE_EXT_TO_DECODER, ext)) {
+                const decoderConstructor = FILE_EXT_TO_DECODER[ext];
+                if (filename.endsWith(ext) && "undefined" !== typeof decoderConstructor) {
+                    return decoderConstructor;
+                }
+            }
+        }
+
+        return null;
+    };
+
+    constructor (pageSize: number) {
+        this.#pageSize = pageSize;
+    }
+
+    async loadFile (fileSrc: FileSrcType): Promise<number> {
+        if ("string" === typeof fileSrc) {
+            this.#fileName = getBasenameFromUrl(fileSrc);
+            this.#fileData = await getUint8ArrayFrom(fileSrc, () => null);
+        } else {
+            // TODO: support file loading via Open / Drag-n-drop
+            console.error("Read from File not yet supported");
+        }
+
+        if (null === this.#fileData || null === this.#fileName) {
+            return 0;
+        }
+
+        const MatchingDecoder = LogFileManager.#getDecoderConstructor(this.#fileName);
+        if (null === MatchingDecoder) {
+            return 0;
+        }
+
+        this.#decoder = new MatchingDecoder(this.#fileData);
+        this.#numEvents = this.#decoder.buildIdx();
+        console.log(`Found ${this.#numEvents} log events.`);
+
+        return this.#numEvents;
+    }
+
+    setDecodeOptions (options: DecodeOptionsType) {
+        if (null === this.#decoder) {
+            throw new Error("loadFile() must be first called.");
+        }
+        this.#decoder.setDecodeOptions(options);
+    }
+
+    loadPage (cursor: CursorType): {
+        logs: string,
+        lines: LineNumLogEventNumMap,
+        cursorLineNum: number
+    } {
+        console.debug(`loadPage: cursor=${JSON.stringify(cursor)}`);
+        if (null === this.#decoder) {
+            throw new Error("loadFile() must be first called.");
+        }
+
+        const results: Array<[string, number, number, number]> = [];
+        let startLogEventNum = (Math.floor(this.#numEvents / this.#pageSize) * this.#pageSize) + 1;
+        if (null !== cursor) {
+            if ("pageNum" in cursor) {
```

Comment:
I haven't thought this through, but rather than type-checking like this, could we do something like what you did for the request/response? I.e., each cursor type has a type code and the arguments for the cursor type depend on the type code.

Reply:
Right, the type code is required for the generic function to work. Let me look into that.